keps/sig-apps/20190318-PodDisruptionBudget-graduation-to-stable.md

```diff
@@ -0,0 +1,134 @@
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+[Pod Disruption Budget (PDB)](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)
+is a Kubernetes API that limits the number of pods of a collection that are down simultaneously from voluntary disruptions.
+[Kubernetes eviction API](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#the-eviction-api)
+takes PDB into account when terminating pods. If PDB is
+violated, the eviction API returns failure and does not delete the requested pod.
+The feature was introduced in Kubernetes 1.4 and promoted to beta in 1.5.
+It has been in beta for a long time. This document lays out the plan to promote
+it to stable.
+
+## Motivation
+
+PDB API has been stable and is an important feature that allows users to improve
+reliability of their critical workloads. This feature has been in beta for a
+long time. We need to promote to stable version given that we plan to support it
+long term.
+
+### Goals
+
+* Plan to promote PDB API to stable version.
+* List action items to mitigate potential risks of supporting mutable PDB.
+
+### Non-Goals
+
+* Making changes to the API fields or their meanings.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### Mutable PDBs
+
+A mutable PDB object allows its `MinAvailable`, `MaxUnavailable`, and `Selector`
+fields to be modified by clients. Components that use PDB must watch such
+modifications and use the updated values when making decisions.
+
+This feature is implemented by [this PR](https://github.com/kubernetes/kubernetes/pull/69867).
+
+### Risks and Mitigations
+
+We plan to support mutation of PDB objects that didn't exist in previous versions.
+The following needs to be checked and verified before graduation of the API.
+
+- [ ] Ensure that components do not have any logic that relies on immutability
+of PDBs. For example, if component builds a cache of pods that match various
+PDBs, they must add logic to invalid the cache on updates.
+   - Action Item: sweep for in-tree (including kubernetes/* repos) uses to make
+   sure components are informer driven and aren’t assuming immutability.
+- [ ] Check what watches PDBs and make sure there is no performance concerns.
+- [ ] Updating a PDB could cause the state of a cluster to seem incorrect. For
+example, a PDB states that at least 10 replicas of a collection must be alive.
+Kubernetes control plane evicts some of the existing pods, but keeps at least 10
+around. The PDB is updated and states that at least 20 replicas must be kept
+alive. It may appear to an observer that the evictions happened before the PDB 
+update were incorrect, if they don’t notice the PDB update.
```

Comment:
We already have this problem before PDBs are mutable. 

Immutable PDBs can be deleted and re-created with new spec (20 replicas in this example).